src/messages.rs

```diff
@@ -106,36 +106,45 @@ pub struct Message {
     pub(crate) unverified_bytes: Vec<u8>,
 }
 
-/// Manually implement Debug instance to avoid printing unverified bytes.
-impl Debug for Message {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("Message")
-            .field("message_type", &self.message_type)
-            .field("session identifier", &self.identifier)
-            .field("from", &self.from)
-            .field("to", &self.to)
-            .finish()
-    }
-}
-
 impl Message {
     /// Creates a new instance of [`Message`].
     #[instrument(skip_all)]
-    pub(crate) fn new(
+    pub fn new<T>(
         message_type: MessageType,
         identifier: Identifier,
         from: ParticipantIdentifier,
         to: ParticipantIdentifier,
-        unverified_bytes: &[u8],
-    ) -> Self {
+        unverified_bytes: &T,
+    ) -> Result<Self>
+    where
+        T: Serialize,
+    {
         trace!("New message created.");
-        Self {
+        Ok(Self {
             message_type,
             identifier,
             from,
             to,
-            unverified_bytes: unverified_bytes.to_vec(),
-        }
+            unverified_bytes: serialize!(unverified_bytes)?,
+        })
+    }
+
+    /// Creates a new instance of [`Message`] from serialized data.
+    pub fn new_from_serialized_data(
+        message_type: MessageType,
+        identifier: Identifier,
+        from: ParticipantIdentifier,
+        to: ParticipantIdentifier,
+        unverified_bytes: Vec<u8>,
+    ) -> Result<Self> {
+        trace!("New message created.");
```

Comment:
suggestion: you might leave a different tracing statement in the two new functions so it's easier to differentiate which one was called in the logs.